server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java

```diff
@@ -592,8 +717,13 @@ public QueryDataSource query(SingleSeriesExpression seriesExpression, QueryConte
    */
   public int countUpgradeFiles() {
     int totalUpgradeFileNum = 0;
-    for (StorageGroupProcessor storageGroupProcessor : processorMap.values()) {
-      totalUpgradeFileNum += storageGroupProcessor.countUpgradeFiles();
+    for (VirtualStorageGroupManager virtualStorageGroupManager : processorMap.values()) {
+      for (StorageGroupProcessor storageGroupProcessor : virtualStorageGroupManager
+          .getAllVirutalStorageGroupProcessor()) {
+        if (storageGroupProcessor != null) {
+          totalUpgradeFileNum += storageGroupProcessor.countUpgradeFiles();
+        }
+      }
```

Comment:
Optional: move such loops into VirtualSGManager to reduce upper-level changes.

Reply:
fixed~